std-bits/base/src/main/java/org/enso/base/FileLineReader.java

```diff
@@ -0,0 +1,402 @@
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.enso.base.arrays.LongArrayList;
+import org.graalvm.polyglot.Context;
+
+public class FileLineReader {
+  public static class ByteArrayOutputStreamWithContains extends ByteArrayOutputStream {
+    public ByteArrayOutputStreamWithContains(int size) {
+      super(size);
+    }
+
+    // ** Creates a preloaded stream from a byte array. */
+    public static ByteArrayOutputStreamWithContains fromByteArray(byte[] bytes) {
+      var stream = new ByteArrayOutputStreamWithContains(0);
+      stream.buf = bytes;
+      stream.count = bytes.length;
+      return stream;
+    }
+
+    // ** Based on https://www.geeksforgeeks.org/kmp-algorithm-for-pattern-searching/ */
+    public static int[] computeLongestPrefix(byte[] bytes) {
+      int[] longestPrefix = new int[bytes.length];
+
+      int i = 1;
+      int len = 0;
+      while (i < bytes.length) {
+        if (bytes[i] == bytes[len]) {
+          len++;
+          longestPrefix[i++] = len;
+        } else if (len == 0) {
+          longestPrefix[i++] = 0;
+        } else {
+          len = longestPrefix[len - 1];
+        }
+      }
+
+      return longestPrefix;
+    }
+
+    public boolean contains(byte[] bytes, int[] longestPrefix) {
+      if (bytes.length > count) {
+        return false;
+      }
+
+      int i = 0;
+      int j = 0;
+      while ((count - i) >= (bytes.length - j)) {
+        if (buf[i] == bytes[j]) {
+          i++;
+          j++;
+        }
+
+        if (j == bytes.length) {
+          return true;
+        }
+
+        if (i < count && buf[i] != bytes[j]) {
+          if (j != 0) {
+            j = longestPrefix[j - 1];
+          } else {
+            i++;
+          }
+        }
+      }
+
+      return false;
+    }
+  }
+
+  private static class CancellationToken {
+    public boolean isCancelled = false;
+
+    public void cancel() {
+      isCancelled = true;
+    }
+  }
+
+  private static final Logger LOGGER = Logger.getLogger("enso-file-line-reader");
+
+  // ** Amount of data to read at a time for a single line (4KB). */
+  private static final int LINE_BUFFER = 4 * 1024;
+
+  // ** Amount of data to read at a time (4MB). */
+  private static final int BUFFER_SIZE = 4 * 1024 * 1024;
+
+  private static boolean moreToRead(int c, MappedByteBuffer buffer) {
+    return switch (c) {
+      case '\n', -1 -> false;
+      case '\r' -> {
+        c = buffer.hasRemaining() ? buffer.get() : '\n';
+        if (c != '\n') {
+          buffer.position(buffer.position() - 1);
+        }
+        yield false;
+      }
+      default -> true;
+    };
+  }
```

Comment:
What does this function check for? I'm not sure if I understand. Does it check if there is more data _in the current line_?